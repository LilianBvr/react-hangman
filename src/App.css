.hangman{

	margin: 3vh 0 3vh 0; padding: 0;
	height: 94vh;

	display: flex;
	flex-wrap: nowrap;
	flex-direction: column;
	justify-content:space-around;
	align-items: center;

	font-family: sans-serif;
	background-color: #DDDDDDDD;

	user-select: none;
}

.hangman .restart{
	margin:5px;
	min-width: 40px;min-height: 40px;

	background-color: #DDDDDDDD;
  	border: 2px solid #888;
  	border-radius: 4px;

  	font-family:arial;
  	font-size:1.2em;
  	color: #888;
  	text-align: center;
}

.hangman > .restart:hover{
	background-color:  #888;
	color:white;
	transition: all 0.3s;
	cursor: pointer;
}
