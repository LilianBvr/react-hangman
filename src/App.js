import React, {Component} from 'react';

import './App.css';

import Mask from './Mask'
import Keyboard from './Keyboard'

export const LETTERS = Array.from('ABCDEFGHIJKLMNOPQRSTUVWXYZ')

class App extends Component {

	constructor(){
		super()

		/*	Etat local	*/
		this.state = this.getInitState()
	}
	
	/*	Méthode de rendu du composant	*/
	render(){
		const { mask, testedLetters, won } = this.state
	  	return (
	    	<div className="hangman">
	    		<Mask mask={mask} />

	    		{
	    			won ? <button className="restart" onClick={this.handleRestartClick}>
							Recommencer
						</button> :
	    				<Keyboard 
	    					letters={LETTERS} 
	    					testedLetters={testedLetters}
	    					onClick={this.handleLetterClick}
	    				/>
	    		}

	    	</div>
	  	)
	}

	/*	Gestionnaire de clic sur une lettre 	*/
	/* @autobind */
	handleLetterClick = letter => {
		const { mask, testedLetters, word } = this.state
		const newMask = this.revealLetter(letter)

		if(!testedLetters.includes(letter))
		{
			testedLetters.push(letter)

			this.setState({
				testedLetters: testedLetters,
				mask: newMask,
				won: word.every((char, index) => char === newMask[index] )
			})
		}
	}

	/*	Gestionnaire de l'évènement Recommencer*/
	/* @autobind */
	handleRestartClick = () => { 
		this.setState( this.getInitState() ) 
	}

	/*	Révèle la lettre en paramètres et retourne le nouveau masque	*/
	revealLetter( letter ){
		const { mask, word } = this.state

		return mask.map((char, i) => letter===word[i] ? word[i] : char)
	}

	/*	Renvoie un mot aléatoire de la liste locale	*/
	getRandomWord(){
		return Array.from(WORDS[ Math.floor(Math.random() * WORDS.length )]);
	}

	/*	Renvoie l'état initial	*/
	getInitState(){
		const word = this.getRandomWord()

		return { 
			word: word,
			mask: this.getInitialMask(word),
			testedLetters: [],
			won: false
		}
	}

	/*	Méthode retournant le mot masqué de départ	*/
	getInitialMask(word){
		return word.map((letter, index)=>[0,word.length-1].includes(index) ? letter : '_')
	}
}


const WORDS = [
	'MANGER',
	'BOIRE',
	'DORMIR',
	'POMME',
	'POIRE',
	'FRAISE',
	'FRAMBOISE',
	'SUPREME',
	'PALACE',
	'GUCCI'
]


export default App;
