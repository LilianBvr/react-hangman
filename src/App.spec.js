import React from 'react'
import { mount, shallow } from 'enzyme'
import { expect } from 'chai'
import sinon from 'sinon'

import App from './App'


describe('<App />', () => {
	
	it('should render and match its reference snapshot', () => {
		const mock = sinon
			.stub( App.prototype, 'getRandomWord')
			.returns(Array.from('TEST'))

		try{
			const wrapper = shallow(<App />)
			expect(wrapper).to.matchSnapshot()
		}
		finally{
			mock.restore()
		}
	})
})