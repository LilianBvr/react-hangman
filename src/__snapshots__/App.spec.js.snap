// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render and match its reference snapshot 1`] = `
<div
  className="hangman"
>
  <Mask
    mask={
      Array [
        "T",
        "_",
        "_",
        "T",
      ]
    }
  />
  <Keyboard
    letters={
      Array [
        "A",
        "B",
        "C",
        "D",
        "E",
        "F",
        "G",
        "H",
        "I",
        "J",
        "K",
        "L",
        "M",
        "N",
        "O",
        "P",
        "Q",
        "R",
        "S",
        "T",
        "U",
        "V",
        "W",
        "X",
        "Y",
        "Z",
      ]
    }
    onClick={[Function]}
    testedLetters={Array []}
  />
</div>
`;
